### PR TITLE
Generate documentation from associations

### PR DESCRIPTION
**Pull request by stevepm**
- Associations supported are `has_many`, `has_one`, `belongs_to` as well as `polymorphic` associations

Signed-off-by: Dor Dan <ddan@mavenlink.com>

### CODE REVIEW

File: lib/brainstem/api_docs/formatters/open_api_specification/version_2/endpoint/response_definitions_formatter.rb [hunk 1]
```diff
@@ -73,41 +73,71 @@ def success_response_description
...
+                    type: 'array',
+                    items: {
+                      type: 'object',
+                      properties: {
+                        key: type_and_format('string'),
+                        id: type_and_format('string')
                       }
                     }
+                  },
+                  brainstem_key => {
```
Comment (shirish-pampoorickal) [inline lib/brainstem/api_docs/formatters/open_api_specification/version_2/endpoint/response_definitions_formatter.rb]: I think you can pull this into a method where you pass in the key and target_klass

Comment (stevepm) [reply lib/brainstem/api_docs/formatters/open_api_specification/version_2/endpoint/response_definitions_formatter.rb]: 👍

File: lib/brainstem/api_docs/formatters/open_api_specification/version_2/presenter_formatter.rb [hunk 2]
```diff
@@ -52,18 +52,45 @@ def format_type!
...
               end
             end
-            
+
+            def format_field_associations(properties)
+              presenter.valid_associations.each_with_object(properties) do |(_name, association), props|
+                if association.foreign_key
+                  case association.type
+                  when :belongs_to, :has_one
+                    props[association.foreign_key] = type_and_format(:integer).merge(description: association.description) unless props[association.foreign_key]
```
Comment (shirish-pampoorickal) [inline lib/brainstem/api_docs/formatters/open_api_specification/version_2/presenter_formatter.rb]: The description should not use the association description but we should add a description stating that this field only shows up the association is included.

File: lib/brainstem/api_docs/formatters/open_api_specification/version_2/presenter_formatter.rb [hunk 3]
```diff
@@ -52,18 +52,45 @@ def format_type!
...
+              branch.each_with_object(ActiveSupport::HashWithIndifferentAccess.new) do |(name, field), buffer|
                 buffer[name.to_s] = format_field(field)
-                buffer
               end
             end
-            
+
+            def format_field_associations(properties)
+              presenter.valid_associations.each_with_object(properties) do |(_name, association), props|
+                if association.foreign_key
```
Comment (shirish-pampoorickal) [inline lib/brainstem/api_docs/formatters/open_api_specification/version_2/presenter_formatter.rb]: We should check if the foreign key field already exists then ignore this attribute. Example, a story has a persistent field workspace_id & is always included. If we do not check for the existence of the attribute, we will overwrite it.

File: lib/brainstem/api_docs/formatters/open_api_specification/version_2/presenter_formatter.rb [hunk 4]
```diff
@@ -52,18 +52,45 @@ def format_type!
...
+                  end
+                end
+
+                if association.polymorphic?
+                  props[association.name + "_ref"] = {
+                    type: 'object',
+                    description: association.description,
+                    properties: {
+                      key: type_and_format(:string),
+                      id: type_and_format(:string)
```
[the remark this answers is not in the document]
Comment (stevepm) [reply lib/brainstem/api_docs/formatters/open_api_specification/version_2/presenter_formatter.rb]: yes

File: lib/brainstem/api_docs/formatters/open_api_specification/version_2/presenter_formatter.rb [hunk 5]
```diff
@@ -52,18 +52,45 @@ def format_type!
...
-            
+
+            def format_field_associations(properties)
+              presenter.valid_associations.each_with_object(properties) do |(_name, association), props|
+                if association.foreign_key
+                  case association.type
+                  when :belongs_to, :has_one
+                    props[association.foreign_key] = type_and_format(:integer).merge(description: association.description) unless props[association.foreign_key]
+                  when :has_many
+                    prop_key = "#{association.foreign_key}s"
```
[the remark this answers is not in the document]
Comment (stevepm) [reply lib/brainstem/api_docs/formatters/open_api_specification/version_2/presenter_formatter.rb]: that works

File: lib/brainstem/api_docs/formatters/open_api_specification/version_2/presenter_formatter.rb (hunk 3 above)
Comment (stevepm) [reply lib/brainstem/api_docs/formatters/open_api_specification/version_2/presenter_formatter.rb]: we're already doing that

[the remark this answers is not in the document]
Comment (stevepm) [reply lib/brainstem/api_docs/formatters/open_api_specification/version_2/presenter_formatter.rb]: we removed the duplicated code, without having to add additional methods, but feel free to push back if you really want that.

File: lib/brainstem/api_docs/formatters/open_api_specification/version_2/endpoint/response_definitions_formatter.rb [hunk 6]
```diff
@@ -73,41 +73,70 @@ def success_response_description
...
                     }
+                  },
+                  brainstem_key => object_reference(model_klass)
+                }.merge(associated_properties)
+              end
+
+              def associated_properties
+                presenter.valid_associations.each_with_object({}) do |(_key, association), obj|
+                  if association.polymorphic?
+                    associations = association.polymorphic_classes || []
```
Comment (shirish-pampoorickal) [inline lib/brainstem/api_docs/formatters/open_api_specification/version_2/endpoint/response_definitions_formatter.rb]: we should rename this to associated_klasses or something.

File: lib/brainstem/api_docs/formatters/open_api_specification/version_2/presenter_formatter.rb [hunk 7]
```diff
@@ -52,18 +52,61 @@ def format_type!
...
+            end
+
+            def association_key(association)
+              key = if association.foreign_key
+                association.foreign_key
+              else
+                "#{association.name.singularize}_id"
+              end
+
+              association.type == :has_many ? "#{key}s" : key
```
Comment (shirish-pampoorickal) [inline lib/brainstem/api_docs/formatters/open_api_specification/version_2/presenter_formatter.rb]: Shouldn't this belong on line 104? I feel its weird to pluralize the foreign key given by the user.

Comment (stevepm) [reply lib/brainstem/api_docs/formatters/open_api_specification/version_2/presenter_formatter.rb]: with a `:has_many`, the user wouldn't be providing a foreign key, this `key` at this point is generated by the `else` block via ` "#{association.name.singularize}_id"`

Comment (stevepm) [reply lib/brainstem/api_docs/formatters/open_api_specification/version_2/presenter_formatter.rb]: 👍

File: lib/brainstem/api_docs/formatters/open_api_specification/version_2/endpoint/response_definitions_formatter.rb (hunk 6 above)
Comment (stevepm) [reply lib/brainstem/api_docs/formatters/open_api_specification/version_2/endpoint/response_definitions_formatter.rb]: 👍

File: lib/brainstem/api_docs/formatters/open_api_specification/version_2/presenter_formatter.rb [hunk 8]
```diff
@@ -52,18 +52,60 @@ def format_type!
...
+                      id: type_and_format(:string)
+                    }
+                  }
+                  next
+                end
+
+                key = association_key(association)
+                description = association_description(key, association.name)
+                formatted_type = if association.type == :has_many
+                  type_and_format(:array, :integer)
```
Comment (shirish-pampoorickal) [inline lib/brainstem/api_docs/formatters/open_api_specification/version_2/presenter_formatter.rb]: CRITICAL: Please update the item_type of the array to `:string` & also the belongs to foreign key to be `:string`

File: spec/brainstem/api_docs/formatters/open_api_specification/version_2/presenter_formatter_spec.rb [hunk 9]
```diff
@@ -146,6 +147,84 @@ module Version2
...
 
+              context 'with associations present' do
+                context 'when association type is belongs_to or has_one' do
+                  before do
+                    presenter_class.associations do
+                      association :task, Task,
+                        type: :belongs_to
+
+                      association :user, User,
+                        foreign_key: :user_id,
```
Comment (shirish-pampoorickal) [inline spec/brainstem/api_docs/formatters/open_api_specification/version_2/presenter_formatter_spec.rb]: Lets rename `foreign_key` to something better. Maybe `association_field_key` or `model_field_key`...

**Comment by shirish-pampoorickal** [comment]
The comments have been addressed in https://github.com/mavenlink/brainstem/pull/120, so LGTM!

**Pull request merged**